Storage/src/SigningHelper.php

```diff
@@ -0,0 +1,578 @@
+<?php
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Auth\CredentialsLoader;
+use Google\Auth\SignBlobInterface;
+use Google\Cloud\Core\Timestamp;
+use Google\Cloud\Storage\Connection\ConnectionInterface;
+
+/**
+ * Provides common methods for signing storage URLs.
+ *
+ * @internal
+ */
+class SigningHelper
+{
+    const DEFAULT_URL_SIGNING_VERSION = 'v2';
+    const DEFAULT_DOWNLOAD_HOST = 'storage.googleapis.com';
+
+    const V4_ALGO_NAME = 'GOOG4-RSA-SHA256';
+    const V4_TIMESTAMP_FORMAT = 'Ymd\THis\Z';
+    const V4_DATESTAMP_FORMAT = 'Ymd';
+
+    /**
+     * Create or fetch a SigningHelper instance.
+     *
+     * @return SigningHelper
+     */
+    public static function getHelper()
+    {
+        static $helper;
+        if (!$helper) {
+            $helper = new static;
+        }
+
+        return $helper;
+    }
+    /**
+     * Get the credentials for use with signing.
+     *
+     * @param ConnectionInterface $connection A Storage connection object.
+     * @param array $options Configuration options.
+     * @return array A list containing a credentials object at index 0 and the
+     *        modified options at index 1.
+     * @throws \RuntimeException If the credentials type is not valid for signing.
+     * @throws \InvalidArgumentException If a keyfile is given and is not valid.
+     */
+    public static function getSigningCredentials(ConnectionInterface $connection, array $options)
```

Comment:
I'm wondering if we could reduce the orchestration required in the consuming classes by just having:

```php
public static function sign(ConnectionInterface $connection, array $options)
// delegates out to v2/v4 signing implementations
```

```php
// consumer
public function signedUrl(...)
{
    ...
    return SigningHelper::sign($this->connection, $options);
}
```

WDYT? Did we have plans to extend this in someway where the additional orchestration makes more sense?